Places fetcher

tbarnahor opened the pull request:
Initial Version of PlacesFetcher, which still doesn't include:
Exception handling
The ability to search more than one type of cuisine
We may want to return some additional details that aren't returned in this version (Places status, Place cuisine types, etc.)

tbarnahor committed:
add placesFetcher

tbarnahor committed:
Merge branch 'master' of https://github.com/googleinterns/Food-Mood into placesFetcher

tbarnahor committed:
change returned type to an immutable list of Places

tbarnahor committed:
add zeroResultsTest

tbarnahor committed:
fix merge conflicts

tbarnahor committed:
add valid search test

tbarnahor committed:
Merge branch 'master' of https://github.com/googleinterns/Food-Mood into placesFetcher

tbarnahor committed:
add placeFetcher (M0 version)

tbarnahor committed:
some small fixes

tbarnahor committed:
Delete wierd files added project

karenav commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,149 @@
+import com.google.maps.errors.ApiException;
+import com.google.maps.model.PlacesSearchResult;
+import com.google.maps.PlacesApi;
+import java.util.*;
```
from what I understand it's not recommended to import the whole package (.*), it's best to explicitly import the components that you need to use.

tbarnahor replied:
You are right :) Changed it

karenav commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public List<Place> fetch() throws IOException, InterruptedException, ApiException { // TODO (talbarnahor): add exception handling and testing
+        PlacesSearchResult results[] = getPlacesSearchResults();
+        System.out.println(results.length);
```
notice the print

tbarnahor replied:
Yeah thanks!

karenav commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,149 @@
+    public PlacesSearchResult[] getPlacesSearchResults()
+            throws ApiException, InterruptedException, IOException {
+        TextSearchRequest query = 
+        PlacesApi.textSearchQuery(CONTEXT, cuisineType, location)
```
I think there needs to be an indentation of 4 spaces at the beginning of the line

tbarnahor replied:
right, I didn't notice it's missing. Fixed it

karenav commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,149 @@
+        List<Place> places = new ArrayList<Place>();
+        for (PlacesSearchResult searchResult: searchResultsArr) {
+            PlaceDetails placeDetails = getPlaceDetails(searchResult.placeId);
+            Place place = Place.create(
```
Once the builder will be pushed, this should be altered to build a new Place (instead of create). You can see an example of how to use the builder in the Place test :)

tbarnahor replied:
Great, I changed it to Build.

karenav commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,116 @@
+  private static final Place PLACE_2 = Place.create("name2", urlString, phone, rating, priceLevelInt, location);
+
+  /** PlacesSearchResults to be used in tests. */
+  private static PlacesSearchResult SEARCH_RESULT_1 = createTestPlacesSearchResult(placeId_1);
```
same thing here, in the future will change to builder :)

karenav commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,116 @@
+      ImmutableList<Place> expectedOutput = ImmutableList.of(PLACE_1, PLACE_2);
+      assertEquals(expectedOutput, spiedFetcher.fetch());        
+    } 
+
```
I think this is a redundant space

tbarnahor replied:
The linter thought so too :)

tbarnahor committed:
some linter fixes

tbarnahor committed:
fix merge conflicts regarding Places builder

tbarnahor committed:
Merge branch 'placesFetcher' of https://github.com/googleinterns/Food-Mood into placesFetcher

tbarnahor committed:
fix linter errors

tbarnahor committed:
more linter fixes

szwbar commented on project/pom.xml:
```diff
@@ -33,6 +33,16 @@
+        <version>0.15.0</version>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
```
What's this dependency? Can you please document?

tbarnahor replied:
It's a dependency on the Simple Logging Facade for Java (SLF4J) for the use of google-maps-services. Added documentation.

szwbar replied:
Ack, thanks. Interesting that they don't pull it in themselves...

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.appengine.repackaged.com.google.common.collect.ImmutableList;
```
Move import statement further up?

tbarnahor replied:
Done

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+
+    /** The entry point for a Google GEO API request. */
+    private static final GeoApiContext CONTEXT = new GeoApiContext.Builder()
+        .apiKey("AIza...") // TODO : save key in a file that can be accessed and pushed to github
```
TODO(M0) ?

tbarnahor replied:
The TODO is now done 👍 
I added the API key as an environment variable by adding 
            "env": {
                "API_KEY": "My API Key"
            }
to the launch.json file in .vscode folder

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+    /** Specifies if the fetched places must be open at the time of fetching. */
+    private final boolean openNow;
+
+    /** The type of places that will be searched. */
```
This comment is not very helpful. Prefer to explain here *why* `RESTAURANT` was chosen as the place to be searched.

tbarnahor replied:
sure

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+
+public class PlacesFetcher {
+
+    /** Places will be fetched within a ceratain radius from this location. */
```
1. No need for JavaDoc on private fields. 
   If this is a linter error, we should remove the error. 
2. Why do you need `private` fields in this class? Do you intend to instantiate a single instance of this class for each request? Does this class need to preserve state along the request? 
    If not, why not just accept the request parameters in the `fetch` method, and then remove these private fields?  
   For the M0 stage, maybe it's better to hold the values statically (as you had before) as constants.

tbarnahor replied:
1. Yeah I added the javaDoc because of the linter errors. I will try to remove the error.
2. I added them as class fields to avoid passing them to the fetch() and from fetch() to getPlacesSearchResults(), but I understand now that it is better to pass them to each function and keep the ability to instantiate a single instance of this class.
I changed the values to be constants for M0 as suggested.

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+        .build();
+
+    /**
+     * Fields are temporaraly hard coded for M0 version.
```
nit: hardcoded (as a single word)

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+     * @throws InterruptedException
+     * @throws IOException
+     */
+    public PlacesSearchResult[] getPlacesSearchResults()
```
Why is this method public? 
It's a helper method for `fetch`, so it should be private (and does not require any Javadoc as a result).

tbarnahor replied:
I changed it to public because otherwise I wasn't able to stub it in the tester. 
Anyway I will try to rearrange it like you suggested on the testing comment and maybe it will solve this visibility issue.

szwbar replied:
OK, let me know when you finish the test change.

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+                .radius(SEARCH_RADIUS)
+                .maxPrice(maxPriceLevel)
+                .type(TYPE);
+        if (openNow) {
```
From the `openNow` documentation in `TextSearchRequest`:
`@param openNow Whether to restrict this search to open places.`

In my understanding this means that if you provide "false" as input, the search is *not* restricted to (but may contain) places that are open now. In this case, I think you shouldn't have the condition here (`if (openNow) ...`).

Then, you can concatenate the call with `await()` which is more readable.

tbarnahor replied:
At first I understood it like you, but when testing I found that changing openNow to false is equivelent to seaching it as true, and only when I ommitted the querry.openNow I recieved results that include closed restaurants. 
You can see examples for this behavor [here on "NYC search examples"](https://docs.google.com/document/d/1D3O6nNO__P5a3gH8zC_TrIkc8ujWpRI7W4piUWYMWew/edit?usp=sharing), including an explanation about it from the official library documentation.

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+        return results.results;
+    }
+
+    /**
```
Javadoc is redundant on private methods.

tbarnahor replied:
Removed it :)

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+     * @throws ApiException
+     */
+    public List<Place> fetch() throws IOException, InterruptedException, ApiException {
+        PlacesSearchResult[] results = getPlacesSearchResults();
```
Suggestion: make this a one-liner
`return createPlacesList(getPlacesSearchResults());`

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+    private List<Place> createPlacesList(PlacesSearchResult[] searchResultsArr)
+            throws ApiException, InterruptedException, IOException {
+        List<Place> places = new ArrayList<Place>();
+        for (PlacesSearchResult searchResult: searchResultsArr) {
```
add whitespace before colon (`:`)

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+                    .setLocation(placeDetails.geometry.location)
+                    .build());
+        }
+        return ImmutableList.copyOf(places);
```
Great!

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+@RunWith(JUnit4.class)
+public final class PlacesFetcherTest {
+
+  /** url for Places used in tests. */
```
nit: In comments we'd write in all caps: URL

Also, I don't believe these fields require documentation (and certainly not a JavaDoc).

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+  private static final String URL_STRING = "https://www.google.com/";
+
+  /** url for PlaceDetails used in tests. */
+  private static final URL URL = createTestURL(URL_STRING);
```
Since you import URL as a type, I believe that having `URL` as a variable name is problematic.

tbarnahor replied:
true

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+    createTestPlaceDetails("name2", URL, PHONE, RATING, PRICELEVEL, LOCATION);
+
+  /**
+   * creates URL from String for PlaceDetails used in tests.
```
Nit: capital letter. Also in other comments. 
Also, there's no need for Javadoc on a private method.

tbarnahor replied:
I ommitted them. Just left a short explanation about createTestURL because I think it's a wierd function so maybe it should be explained.

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+   */
+  private static URL createTestURL(String s) {
+    try {
+      return new URL(s);
```
What happens if you just have 

```java
private static final URL x = new URL(URL_STRING)
```
 Does this fail to compile? 
Otherwise I don't think you need this function at all.

tbarnahor replied:
Yes it fails to compile because of un handled exception. The best solution I found is to wrap it with try/catch in a function but I know it's weird.

szwbar replied:
Ack, thanks.

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+   *
+   * @param name The name of the Place
+   * @param url The url of the Place's website
+   * @param phone The phonenumber of the Place
```
nit: phone number

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+   * @param location The location of the Place
+   * @return The PlacesDetails according to the given params
+   */
+  private static PlaceDetails createTestPlaceDetails(String name, URL url, String phone,
```
nit: add a newline after the first `(` for better formatting: 
```java
private static PlaceDetails createTestPlaceDetails(
    String name, URL url, String phone, float rating, PriceLevel priceLevel, LatLng location) {
```

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+    return placeDetails;
+  }
+
+  /** creates PlaceSearchResult to be used in tests.
```
Newline after /**

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+    return searchResult;
+  }
+
+  /** A PlacesFetcher instance to be spied on. */
```
I think that you more than spy on it - this is the class under test!

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -0,0 +1,157 @@
+     * @throws InterruptedException
+     * @throws ApiException
+     */
+    public List<Place> fetch() throws IOException, InterruptedException, ApiException {
```
You can return an `ImmutableList` as part of your API.

szwbar replied:
?

tbarnahor replied:
Sorry I missed it

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+import java.net.URL;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
```
Please decide whether you `import static` all mockito options (when, mock, etc.) or none.

tbarnahor replied:
I don't quite understand, can you please elaborate on that?
I should import options I am not using?

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+  @Test
+  public void fetch_zeroSearchResults_returnsEmptyList() throws Exception {
+    PlacesFetcher spiedFetcher = spy(placesFetcher);
+    doReturn(new PlacesSearchResult[0]).when(spiedFetcher).getPlacesSearchResults();
```
Interesting solution! 
The problem that I see with the approach you took here, is that you won't be able to test the `getPlacesSearchResult` function. 
Instead, I think that you should have a single method that only calls `request.await()` and returns the value. Then you can mock out this method (and this method only), but still test your other method (the one in which you choose inputs). 
Same goes for the place details (you can even combine them). 
LMK if you want to talk about this f2f.

tbarnahor replied:
I understand and will change it. 
Just one dillema - do you think I should make an interface with a test implementation where those request.await() methods will be public so I can mock them, and production implementaion where they are private? Or it doesn't matter for them to be public?

szwbar replied:
First, make the method package-private (aka default) and not public. 
Then, add an annotation of Guava's `@VisibleForTesting` on this method. This will help users of your class understand that this is not intended to be package-private, but must be because of testing. 
https://stackoverflow.com/questions/6913325/annotation-to-make-a-private-method-public-only-for-test-classes

tbarnahor replied:
Sound good, but this means the test has to be in the same package as PlaceFetcher and it's not how our packages are orginized. So you think I should put them in the same package or just use the annotation but with public visibility?

tbarnahor replied:
I changed the functions like you suggested but I ran into a problem with testing the requests sent to the methods that call request.await() so IMO it's still not good enough. I pushed the best thing I got to and we can discuss this f2f and I explain the problem.

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -0,0 +1,177 @@
+  public void fetch_zeroSearchResults_returnsEmptyList() throws Exception {
+    PlacesFetcher spiedFetcher = spy(placesFetcher);
+    doReturn(new PlacesSearchResult[0]).when(spiedFetcher).getPlacesSearchResults();
+    ImmutableList<Place> expectedOutput = ImmutableList.of();
```
Inline the expectation:
```java
assertEquals(ImmutableList.of(), ...);
```

szwbar reviewed (commented):
Great work! Mostly formatting comments and some suggestions. 

General question: why do you also push the `.class` files?

tbarnahor committed:
handle API_KEY and small PR fixes

tbarnahor committed:
add explanation about the use of  RESTAURANT type

tbarnahor committed:
change PlacesFetcher class fields to be constants

tbarnahor committed:
PR fixes

tbarnahor committed:
remove private methods javadoc

tbarnahor committed:
fix PlacesFetcherTest PR comments

tbarnahor committed:
fix formating based on PR comments

tbarnahor committed:
remove target files from PR

tbarnahor committed:
Remove target file from PR

tbarnahor committed:
Delete checkstyle.xml

tbarnahor committed:
change some Fetcher's functions so test can mock only .await()

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -25,52 +26,39 @@
-     * In next versions those same fields will be the fields of a UserPrefrences
-     * instance passed to the PlacesFetcher constructor by the Servlet.
-     */
     public PlacesFetcher() {
```
This constructor is now unneeded. Please remove it.

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -80,8 +68,7 @@ public PlacesFetcher() {
      * @throws ApiException
      */
     public List<Place> fetch() throws IOException, InterruptedException, ApiException {
```
Continuing a conversation I cannot comment on anymore: 
why not accept the inputs as parameters in your library? and let the caller specify them? 
Otherwise you'll need another change to your library, although it's quite ready.

tbarnahor replied:
It will be like that in M1 (as our [design](https://docs.google.com/document/d/1yPTb2Rh6MSTrDE9b9e1Q61ZCcH_cVNqBI8qs1VKNZEs/edit#) was planned. )
For now since in M0 we said it will all be hardcoded in the PlacesFetcher I don't see how inputs to fetch() are useful.

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -95,35 +82,27 @@ public PlacesFetcher() {
                 Place.builder()
                     .setName(placeDetails.name)
-                    .setWebsiteUrl(placeDetails.website.toString())
+                    .setWebsiteUrl((placeDetails.website == null)
```
nit: redundant parantheses.

szwbar reviewed (changes requested):
What about the extra .class files? Are they needed?

tbarnahor committed:
more PR fixes

tbarnahor committed:
Merge branch 'master' of https://github.com/googleinterns/Food-Mood into placesFetcher

tbarnahor committed:
Merge branch 'placesFetcher' of https://github.com/googleinterns/Food-Mood into placesFetcher

tbarnahor committed:
move binary files from PR

tbarnahor commented:
Regarding the .class files, they got pushed by accident. I removed them from this PR and also I added them to my local gitignore file so this won't happen again (tested it and they are indeed ignored from now on).

tbarnahor committed:
linter fixes

tbarnahor committed:
final linter fix

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -55,55 +55,55 @@
      */
-    public PlacesSearchResult[] getPlacesSearchResults()
+    @VisibleForTesting
+    public PlacesSearchResult[] getPlacesSearchResults(TextSearchRequest query)
```
Since the test is in the same package as `PlacesFetcher`, a package-private modifier should be enough.

szwbar commented on project/src/main/java/com/google/sps/data/PlacesFetcher.java:
```diff
@@ -112,25 +112,29 @@ public PlacesFetcher() {
+     * @throws IOException
+     */
+    @VisibleForTesting
+    public PlaceDetails getPlaceDetails(PlaceDetailsRequest request)
```
Modifier

szwbar commented on project/src/test/java/com/google/sps/test/PlacesFetcherTest.java:
```diff
@@ -119,17 +121,19 @@ private static PlacesSearchResult createTestPlacesSearchResult(String placeId) {
   }
 
   @Test
   public void fetch_validSearchResults_returnsListOfPlaces()
```
This comment is theoretical only:
Now that you have that spy in place (which is really close to an actual mock BTW!), you *could* test that the parameters that you provided are OK. 

You could add new tests that `getPlaceDetails` and  `getPlacesSearchResults` are called with the parameters you expect it to be called. This test your logic when building the request. 
For example, you could test that the given location matches the location you gave the function itself.

*However*, there are no getters on `TextSearchRequest` and `PlaceDetailsRequest`, sadly, so you will not be able to do that.

tbarnahor replied:
Yes, I really tried to do that for a while yesterday because it's bugging me that those methods are not really tested but I didn't find any way to get the info from the requests so I gave up for now.

tbarnahor committed:
change tests  folder so tests and classes are in the same package
